pkg/cloud/services/ec2/launchtemplate.go

```diff
@@ -0,0 +1,323 @@
+			VolumeSize:          aws.Int64(lt.RootVolume.Size),
+			Encrypted:           aws.Bool(lt.RootVolume.Encrypted),
+		}
+
+		if lt.RootVolume.IOPS != 0 {
+			ebsRootDevice.Iops = aws.Int64(lt.RootVolume.IOPS)
+		}
+
+		if lt.RootVolume.EncryptionKey != "" {
+			ebsRootDevice.Encrypted = aws.Bool(true)
+			ebsRootDevice.KmsKeyId = aws.String(lt.RootVolume.EncryptionKey)
+		}
+
+		if lt.RootVolume.Type != "" {
+			ebsRootDevice.VolumeType = aws.String(lt.RootVolume.Type)
+		}
+
+		data.BlockDeviceMappings = []*ec2.LaunchTemplateBlockDeviceMappingRequest{
+			{
+				DeviceName: rootDeviceName,
+				Ebs:        ebsRootDevice,
+			},
+		}
+	}
+
+	return data, nil
+}
+
+// DeleteLaunchTemplate delete a launch template
+func (s *Service) DeleteLaunchTemplate(id string) error {
+	s.scope.V(2).Info("Deleting launch template", "id", id)
+
+	input := &ec2.DeleteLaunchTemplateInput{
+		LaunchTemplateId: aws.String(id),
+	}
+
+	if _, err := s.EC2Client.DeleteLaunchTemplate(input); err != nil {
+		return errors.Wrapf(err, "failed to delete launch template %q", id)
+	}
+
+	s.scope.V(2).Info("Deleted launch template", "id", id)
+	return nil
+}
+
+// SDKToLaunchTemplate converts an AWS EC2 SDK instance to the CAPA instance type.
+func (s *Service) SDKToLaunchTemplate(d *ec2.LaunchTemplateVersion) (*expinfrav1.AWSLaunchTemplate, error) {
+	v := d.LaunchTemplateData
+	i := &expinfrav1.AWSLaunchTemplate{
+		ID:   aws.StringValue(d.LaunchTemplateId),
+		Name: aws.StringValue(d.LaunchTemplateName),
+		AMI: infrav1.AWSResourceReference{
+			ID: v.ImageId,
+		},
+		IamInstanceProfile: aws.StringValue(v.IamInstanceProfile.Name),
+		InstanceType:       aws.StringValue(v.InstanceType),
+		SSHKeyName:         v.KeyName,
+		VersionNumber:      d.VersionNumber,
+	}
+
+	// Extract IAM Instance Profile name from ARN
+	if v.IamInstanceProfile != nil && v.IamInstanceProfile.Arn != nil {
+		split := strings.Split(aws.StringValue(v.IamInstanceProfile.Arn), "instance-profile/")
+		if len(split) > 1 && split[1] != "" {
+			i.IamInstanceProfile = split[1]
+		}
+	}
+
+	for _, ni := range v.NetworkInterfaces {
+		var s []string
+		for _, groups := range ni.Groups {
+			s = append(s, *groups)
+		}
+		tmp := &expinfrav1.NetworkInterface{
+			DeviceIndex: *ni.DeviceIndex,
+			Groups:      s,
+		}
+		i.NetworkInterfaces = append(i.NetworkInterfaces, *tmp)
+	}
+
+	for _, id := range v.SecurityGroupIds {
+		// This will include the core security groups as well, making the "Additional" a bit
+		// dishonest. However, including the core groups drastically simplifies comparison with
+		// the incoming security groups.
+		i.AdditionalSecurityGroups = append(i.AdditionalSecurityGroups, infrav1.AWSResourceReference{ID: id})
+	}
+
+	return i, nil
+}
+
+// LaunchTemplateNeedsUpdate checks if a new launch template version is needed
+func (s *Service) LaunchTemplateNeedsUpdate(incoming *expinfrav1.AWSLaunchTemplate, existing *expinfrav1.AWSLaunchTemplate) (bool, error) {
```

Comment:
This function doesn't currently account for changes in AMI ID, which is calculated in `s.createLaunchTemplateData`. We may want to consider looking up the AMI ID before entering this function, or perhaps adding a separate bool function like "s.NewAMIDiscovered" to help us out here.

Reply:
ok i attempted to address this